{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Answers\n",
    "\n",
    "This pdf document contains solutions for questions asked as part of the \"Genome Annotation\" module."
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Answers for 'Comparing Reference Genomes':\n",
    "\n",
    "**Question: Based on your work during the previous assembly module, can you think of a reason why assembly might not be perfect?**\n",
    "\n",
    "There are many possible answers here (refer to slides 3 & 4 from the assembly presentation), including:\n",
    "\n",
    "1. Repetitive sequences.\n",
    "2. Low sequencing depth.\n",
    "3. Errors in reads\n",
    "\n",
    "**Question: Is there an obvious issue with our assembly?**\n",
    "\n",
    "Yes, the 3' end of our assembly is missing!\n",
    "\n",
    "**Question: Why do you think both ends of the reference genome align to the same part of our assembled genome?**\n",
    "\n",
    "Both ends of a chromosome have telomeres, which contain identical sequence.\n",
    "\n",
    "**Questions: What do you think the green segments represent in this image?**\n",
    "\n",
    "Inversions – pieces of the genome that are in the opposite direction when compared between isolates.\n",
    "\n",
    "**Question: Why is the red line not centered in the plot and moves up or down?**\n",
    "\n",
    "Some pieces of the genome are missing from one isolate to another, meaning the align at different places."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Answers for 'Identifying Repetitive DNA':\n",
    "\n",
    "Below are the repetitive sequences in <t style=\"color:red\"><b>BOLD RED</b></t>:\n",
    "\n",
    "![](images/solutions.repeats.png)\n",
    "\n",
    "**Question: Why do the sections “Retroelements” and “DNA transposons” all have zeros?**\n",
    "\n",
    "_Plasmodium_ species do not have transposons.\n",
    "\n",
    "**Question: Approximately what proportion of our genome assembly was masked?**\n",
    "\n",
    "We can look at the file `PB.contigs.polished.reheader.fasta.tbl` using a command like `less`, `more`, or `cat`. You\n",
    "should see a table similar to the following image:\n",
    "\n",
    "![](images/solutions.rm_output.png)\n",
    "\n",
    "From here you can see that ~17% of the genome was masked (Simple repeats + Low complexity).\n",
    "\n",
    "## Answers for 'Finding Genes':\n",
    "\n",
    "**Question: How many exons does the gene “1_g” have?**\n",
    "\n",
    "Looking at the output of the `head` command you can see that '1_g' has *4* exons. To get this answer, you can also try\n",
    "the more expert-level command:\n",
    "\n",
    "```\n",
    "grep \"\\\"1_g\\\"\" genemark.gtf | grep exon | wc -l\n",
    "```\n",
    "\n",
    "This command should print `4` on the command line. The two `\\\"` in the command above tell `grep` to look for quotation\n",
    "marks as part of the search. The `\\` is what is known as an \"escape character\".\n",
    "\n",
    "**Question: Can you think of a simple LINUX command to figure out how many genes GeneMark-ES identified?**\n",
    "\n",
    "There are multiple solutions to this question – ask your instructor if how you found the solution is acceptable. One\n",
    "possible solution is to use the tool `awk` like below:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "awk '{print $10}' genemark.gtf | uniq | wc -l"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% bash\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command works by:\n",
    "\n",
    "1. Printing the 10th column (`awk '{print $10}'`; \"$10\" is the column with gene_id).\n",
    "2. *PIPING* the output of `awk` to the next command, `uniq`\n",
    "3. `uniq` finds all _unique_ values\n",
    "4. *PIPING* the output of `uniq` to the next command, `wc`\n",
    "5. `wc -l` tells you how many lines are printed by the previous command.\n",
    "\n",
    "To see each step in this command, try running each of the above without the next pipe, like so:\n",
    "\n",
    "1. `awk '{print $10}' genemark.gtf`\n",
    "2. `awk '{print $10}' genemark.gtf | uniq`\n",
    "3. `awk '{print $10}' genemark.gtf | uniq | wc -l`\n",
    "\n",
    "**Question: How many genes did GeneMark find?**\n",
    "\n",
    "You can use the above command to find out. The answer should be close to 300 Genes.\n",
    "\n",
    "**Question: What is the part of the command 2> /dev/null actually doing?**\n",
    "\n",
    "Directing (also known as \"piping\") anything to `/dev/null` means you are directing any output to an empty file that is\n",
    "not actually record\n",
    "\n",
    "**Question: How many genes are in our final set of possible genes (bonafide.gb)?**\n",
    "\n",
    "This requires you to come up with another command by using some abstract problem-solving. If you look at the bonafide.gb\n",
    "file like so:\n",
    "\n",
    "```\n",
    "more bonafide.gb\n",
    "```\n",
    "\n",
    "You may notice that for each gene record, there is a single \"LOCUS\" line. We can use that to find out how many genes\n",
    "were found:\n",
    "\n",
    "```\n",
    "grep -c LOCUS bonafide.gb\n",
    "```\n",
    "\n",
    "the `-c` flag counts the number of times LOCUS was found within the file `bonafide.gb`\n",
    "\n",
    "**Question: What do you think a limitation of using just 1 chromosome to train our gene finder is?**\n",
    "\n",
    "The main limitation is the number of genes that we can use for training. As we are only using 1 chromosome to train\n",
    "AUGUSTUS, there are only ~300 genes that can be used for training. If we use all chromosomes, that means we can use all\n",
    "possible genes in the entire genome (~5,300 genes).\n",
    "\n",
    "**Question: Can you figure out how many genes each approach found?**\n",
    "\n",
    "We can use a similar command as the previous question where we found the number of genes in `bonafide.gb`:\n",
    "\n",
    "```\n",
    "grep -c 'start gene' PB.contigs.default.gff\n",
    "```\n",
    "\n",
    "Since each gff file contains one 'start gene' entry per found gene, we can use `grep` to count the total number of\n",
    "identified genes. This should give a value of approximately 300-320 genes for each file.\n",
    "\n",
    "**Question: We identified protein coding genes, but can you think of any other types of annotations we\n",
    "could find with Augustus?**\n",
    "\n",
    "If you read the AUGUSTUS paper, you can also identify: UnTranslated Regions (UTRs) and Processed pseudogenes.\n",
    "\n",
    "**Question: How many exons does this gene have?**\n",
    "\n",
    "See the below figure:\n",
    "\n",
    "![](images/solutions.gene_one.png)\n",
    "\n",
    "This gene has 4 exons.\n",
    "\n",
    "**Question: How do the predictions for your model versus the default model compare?**\n",
    "\n",
    "The default model uses an alternative splice site for one of the exons.\n",
    "\n",
    "**Question: What are both predictions missing, and why do you think that is?**\n",
    "\n",
    "They are both missing UTRs. That is because we did not tell AUGUSTUS to search for them.\n",
    "\n",
    "**Question: How many exons does this gene have?**\n",
    "\n",
    "It has 8 exons.\n",
    "\n",
    "**Question: How many introns?**\n",
    "\n",
    "It has 7 introns.\n",
    "\n",
    "**Question: How can you extract the same information for another gene by modifying the above command? Report the command\n",
    "and result here.**\n",
    "\n",
    "You can just change \"g54\" to any other gene number and then modify the number given to -A to change the number of lines.\n",
    "An example for another gene (g53) is:\n",
    "\n",
    "```\n",
    "grep -A 15 \"start gene g53\" PB.contigs.gff\n",
    "```"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Answers for 'Using Comparative Genomes to Identify Genes':\n",
    "\n",
    "**Question: What is difficult about this alignment?**\n",
    "\n",
    "It has an intron.\n",
    "\n",
    "**Question: Did you notice something at the end of the alignment that was not in the protein sequence?**\n",
    "\n",
    "The stop codon \"TAG\".\n",
    "\n",
    "**Question: What was difficult in this example?**\n",
    "\n",
    "There is a mutation in the DNA sequence that caused a different amino acid to be in the sequence. The sequence\n",
    "changed from GCA to GGA.\n",
    "\n",
    "**Question: Do you think this is an issue, or is there something biology-related going on?**\n",
    "\n",
    "Changes in DNA sequence occur all the time between species. The change in amino acid could reflect some selective\n",
    "advantage for a given species or be neutral and confer no evolutionary advantage.\n",
    "\n",
    "**Question: What do you think the * character represents?**\n",
    "\n",
    "It is the symbol the stop codon.\n",
    "\n",
    "**Question: What is the gene that we identified in IGV?**\n",
    "\n",
    "Myosin B\n",
    "\n",
    "**Question: Can you name a function of this gene and how did you get the answer?**\n",
    "\n",
    "Myosins are motor proteins that are used for motility via binding of actin chains in eukaryotes. There are a number of\n",
    "resources where this information can be found, including the UCSC genome browser, Uniprot, and _even_ Wikipedia – ;D.\n"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}